### ISSUE TITLE
proposal: `no_lazy_instantiation_of_disposables`

### ISSUE THREAD

**Issue by dnfield** [dart-lang/linter#3058]
## Description

A `late final DisposableFoo foo = DisposableFoo()..doSomething();` will only construct and `doSomething` on first access.

However, if that first access is to call `foo.dispose()`, this is a mistake that should be avoided.

We would probably want some kind of annotation to indicate whether a variable meets this criteria, perhaps with a lint suggestion to so annotate classes that have a `dispose` or `close` method.

## Good Examples 

```dart
class _BarState extends State<Bar> with SingleTickerProviderStateMixin {
  AnimationController<double>? _controller; // initialize elsewhere, e.g. in initState.

  void dispose() {
    _controller?.dispose();
    _controller = null;
  }
}
```

## Bad Examples

```dart
class _BarState extends State<Bar> with SingleTickerProviderStateMixin {
  late final AnimationController<double> _controller = AnimationController<double>(vsync: this, duration: ...)..start();

  void dispose() {
    _controller.dispose(); // Oops. This might be the first access, which is now doing work just to dispose it!
  }
}
```

see https://github.com/flutter/flutter/pull/93166#discussion_r745098688

**Comment by bwilkerson** [comment]
Interesting suggestion.

I like the idea of using an annotation, but I would propose associating the annotation with the specific `close` or `dispose` method(s). That way we don't effectively have a heuristic for finding the methods that fit this pattern.

I was originally going to propose an annotation like `endOfLife` or `invalidating` that would signal that invoking the member implies that the receiver can't be accessed afterward. But I'm not sure whether that's quite the right semantics here. While I'm fairly certain that any controller that's been disposed should not be accessed, it might not be the case that any `endOfLife` method should incur a diagnostic if the receiver is a `late final` field (though I can't think of a concrete example of when that wouldn't be the case).

If we had such an annotation we could also use it to flag any invocations of members after an `endOfLife` member had been invoked, and possibly a lint that requires an `endOfLife` member to be invoked in the same function in which the receiver was instantiated.

There might be other life cycle patterns that we'd also want to consider supporting.

**Comment by dnfield** [comment]
The end of life suggesting is definitely interesting, but it seems like it'd be harder to deal with, e.g.

```
foo.dispose();
// in some other method, called later...
foo = Foo();
```
This is less about tracking lifecycle, and more about making sure you don't lazily instantiate some object only to dispose of it immediately.

Having an annotation on specific methods rather than on classes would be fine.

**Comment by pq** [comment]
Interesting proposal!

I'm curious if the issue isn't more generally about unused disposables than about lazy initialization.

Assuming your "Good" example as written, wouldn't it be just as bad if the controller was only allocated in `initState` and not used?

Or, what about?

```dart
class _BarState extends State<Bar> with SingleTickerProviderStateMixin {
  // Allocated but whoops, never used
  final AnimationController<double>? _controller = AnimationController<double>(vsync: this, duration: ...)..start();

  void dispose() {
    _controller?.dispose();
    _controller = null;
  }
}
```

**Comment by dnfield** [comment]
If you get to `initState`, you're in the tree and it gets a lot harder to tell if you'll be used or not, but it's generally a good assumption thta you will be.

The second example there is a bit different, because it's harder to tell if some other method in the class ended up using the object. At any rate, we instantiated it before disposing it. But with late final we might have instantiated it just to dispose it.
